### PR TITLE
[performance] refactoring + add fave / follow / request / visibility caching

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
There are quite a few changes in this PR! All very welcome imo.
- Refactor the visibility filter 
- Add caching to the visibility filter, including the *many* necessary invalidation callbacks
- Refactor the `bundb.relationship{}` area of the database
- Add caching to follows, follow requests and status favourites, relying on these whenever possible
- Catch more `db.GetWhere()` in the codebase and replace with cache-reliant functions
- Remove unnecessary interfacing of caches
- Allows disabling cache sweeping by setting a sweep frequency of 0
- Update cache configuration defaults to be less conservative with memory usage (previously we were being overly so)
- add support for fetching barebones versions of database models. use these for filling out sub-models

Still missing:
- [x] rebase from upstream/main
- [x] update visibility filter tests (as logic has changed _somewhat_) and potentially add new ones
- [x] ~~add tests for the PopulateStatus() and PopulateAccount() db functions~~
- [x] add tests for new `bundb.relationship{}` functions
- [x] add tests for new `GetAccountBy___()` functions
- [x] add follow / follow request config to the example config yaml
- [x] update env parsing test
- [x] fill out a finished fleshed-out PR description here

Checklist:
- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: testrig/db.go [hunk 1]
```diff
@@ -282,7 +282,7 @@ func StandardDBSetup(db db.DB, accounts map[string]*gtsmodel.Account) {
 func StandardDBTeardown(db db.DB) {
 	ctx := context.Background()
 	if db == nil {
-		log.Panic(nil, "db was nil")
+		return
```
Comment (NyaaaWhatsUpDoc) [inline testrig/db.go:285]: note to kim: undo this change

**Review by tsmethurst — COMMENTED** [review]
Looking good so far, very necessary refactor :heart_eyes:

File: internal/visibility/status.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	if !visible {
+		log.Trace(ctx, "status author not visible to requester")
+		return false, nil
+	}
+
+	if status.BoostOfID != "" {
+		// This is a boosted status.
+
+		if status.AccountID == status.BoostOfAccountID {
+			// Some clout-chaser boosted their own status, tch.
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/status.go:200]: hehe

File: internal/visibility/public_timeline.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	if !visible {
+		log.Trace(ctx, "status not visible to timeline requester")
+		return false, nil
+	}
+
+	for parent := status; parent.InReplyToURI != ""; {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/public_timeline.go:92]: Not even when it's a single author thread of statuses with no mentions?? That is the only case that is really checking for.

File: internal/visibility/filter.go [hunk 4]
```diff
@@ -18,46 +18,20 @@
...
-}
+// noauth is a placeholder ID used in cache lookups
+// when there is no authorized account ID to use.
+const noauth = "noauth"
 
-type filter struct {
-	db db.DB
+// Filter packages up a bunch of logic for checking whether
+// given statuses or accounts are visible to a requester.
+type Filter struct {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/filter.go:30]: :D

File: internal/cache/cache.go [hunk 5]
```diff
@@ -17,43 +17,101 @@
...
 func (c *Caches) Stop() {
 	c.GTS.Stop()
 	c.AP.Stop()
+	c.Visibility.Stop()
+}
+
+// setuphooks sets necessary cache invalidation hooks between caches,
+// as an invalidation indicates a database UPDATE / DELETE. INSERT is
+// not handled by invalidation hooks and must be invalidated manually.
+func (c *Caches) setuphooks() {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/cache/cache.go:70]: :D

File: internal/db/relationship.go [hunk 6]
```diff
@@ -25,42 +25,86 @@ import (
 
 // Relationship contains functions for getting or modifying the relationship between two accounts.
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/relationship.go:26]: Basically when I make new functions I've been using just `error`. It's my intention to get around to removing the `Error` type at some point as we're not actually using that separate type.

File: internal/gtsmodel/account.go [hunk 7]
```diff
@@ -70,8 +71,8 @@ type Account struct {
...
-	PrivateKey              *rsa.PrivateKey  `validate:"required_without=Domain"`                                                                                    // Privatekey for validating activitypub requests, will only be defined for local accounts
-	PublicKey               *rsa.PublicKey   `validate:"required"`                                                                                                   // Publickey for encoding activitypub requests, will be defined for both local and remote accounts
+	PrivateKey              *rsa.PrivateKey  `validate:"required_without=Domain" bun:""`                                                                             // Privatekey for validating activitypub requests, will only be defined for local accounts
+	PublicKey               *rsa.PublicKey   `validate:"required" bun:",notnull,unique"`                                                                             // Publickey for encoding activitypub requests, will be defined for both local and remote accounts
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/gtsmodel/account.go:75]: I mean should a server be sharing keys for accounts? o.O

File: internal/visibility/account.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+		return false, fmt.Errorf("isAccountVisibleTo: error checking account %s visibility: %w", account.ID, err)
+	}
+
+	if !visible {
+		log.Trace(ctx, "requesting account cannot see other accounts")
+		return false, nil
+	}
+
+	// Check whether either blocks the other.
+	blocked, err := f.state.DB.IsEitherBlocked(ctx,
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/account.go:93]: Oh that is a good point... Though I think it will get very tricky returning that specifically without getting a bit of a weird return type like `(bool, bool, error)`. How do you feel about me adding a note about this to the `AccountVisible()` function?

File: internal/db/bundb/relationship_follow_req.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+	r.state.Caches.Visibility.Invalidate("RequesterID", follow.AccountID)
+
+	// Invalidate follow request target account ID cached visibility.
+	r.state.Caches.Visibility.Invalidate("ItemID", follow.TargetAccountID)
+	r.state.Caches.Visibility.Invalidate("RequesterID", follow.TargetAccountID)
+
+	return nil
+}
+
+func (r *relationshipDB) AcceptFollowRequest(ctx context.Context, sourceAccountID string, targetAccountID string) (*gtsmodel.Follow, db.Error) {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/relationship_follow_req.go:170]: The follow request ID gets invalidated below which in turn invalidates the visibility caches through hooks :)

File: internal/visibility/status.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+		log.Trace(ctx, "unauthorized request to non-public status")
+		return false, nil
+	}
+
+	if status.Visibility == gtsmodel.VisibilityUnlocked {
+		// This status is visible to all auth'd accounts.
+		return true, nil
+	}
+
+	if requester.ID == status.AccountID {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/status.go:113]: There is a chance the requester could be nil i.e. when unauthed. Which the above has to account for. And once all the inter-account visibilities have been checked and we have already checked `requester == nil` and `visibility == PUBLIC` then we can do this. We could do a quick check for this above as a "fast path", but considering this result is already being cached I felt that the specific flow of logic was more important. What do you think?

File: internal/db/bundb/notification.go [hunk 11]
```diff
@@ -118,19 +126,37 @@ func (n *notificationDB) DeleteNotification(ctx context.Context, id string) db.E
...
 	// Capture notification IDs in a RETURNING statement.
-	ids := []string{}
+	var ids []string
 
 	q := n.conn.
 		NewDelete().
 		TableExpr("? AS ?", bun.Ident("notifications"), bun.Ident("notification")).
 		Returning("?", bun.Ident("id"))
 
+	if len(types) > 0 {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/notification.go:142]: Oh interesting :eyes:

File: internal/cache/gts.go [hunk 12]
```diff
@@ -25,240 +25,221 @@ import (
...
 
-func (c *gtsCaches) initAccount() {
+func (c *GTSCaches) initAccount() {
 	c.account = result.New([]result.Lookup{
 		{Name: "ID"},
 		{Name: "URI"},
 		{Name: "URL"},
 		{Name: "Username.Domain"},
 		{Name: "PublicKeyURI"},
+		{Name: "InboxURI"},
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/cache/gts.go:208]: :D

File: internal/db/bundb/relationship.go [hunk 13]
```diff
@@ -34,603 +34,210 @@ type relationshipDB struct {
...
-	}
-
-	return *uri, nil
+// newSelectLocalFollows returns a new select query for all rows in the follows table with
+// account_id = accountID where the corresponding account ID has a NULL domain (i.e. is local).
+func newSelectLocalFollows(conn *DBConn, accountID string) *bun.SelectQuery {
+	return conn.NewSelect().
+		Table("follows").
+		Column("id").
+		Where("? = ? AND ? IN ( SELECT ? FROM ? WHERE ? IS NULL )",
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/relationship.go:208]: Honestly I was more confused with how to use the bun API than the SQL itself. I'll double check whether it works with postgres and if it's still all good is it alright if I leave as-is? :')

File: internal/visibility/status.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/visibility/status.go:113]: Aha okay no problem then!

File: internal/cache/visibility.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+// CachedVisibility represents a cached visibility lookup value.
+type CachedVisibility struct {
+	// ItemID is the ID of the item in question (status / account).
+	ItemID string
+
+	// RequesterID is the ID of the requesting account for this visibility lookup.
+	RequesterID string
+
+	// Type is the visibility lookup type: ["status", "account", "home", "public"]
+	Type string
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/cache/visibility.go:64]: Nah that is a good idea! Shall do this.

File: internal/db/relationship.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/relationship.go:26]: What do you think about this?

File: internal/visibility/home_timeline.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	if !visible {
+		log.Trace(ctx, "status not visible to timeline owner")
+		return false, nil
+	}
+
+	if status.AccountID == owner.ID {
```
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/home_timeline.go:87]: This is pretty much the same situation as the check in StatusVisible I was talking about and again this function is more often than not going to be relying on cache. What do you think?

File: internal/visibility/public_timeline.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/visibility/public_timeline.go:92]: oh good point... forgot about that case

in the existing code, we exclude statuses that are a reply in the db query we make to select public statuses (ie., the db query before filtering)

so currently I think (unless you've changed that) this visibility function will never get fed statuses that are replies

i'm OK with changing it if you like, but in that case you should also update the GetPublicTimeline function in `internal/db/bundb/timeline.go`

File: internal/visibility/home_timeline.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/visibility/home_timeline.go:87]: (nvm -- see above)

File: internal/gtsmodel/account.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/gtsmodel/account.go:75]: If there is a legitimate case for it I can remove the unique tag

File: internal/visibility/home_timeline.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/home_timeline.go:87]: The point here is the same as in the `StatusVisible()` comment I made, but specifically for here. What do you think?

File: internal/visibility/account.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/visibility/account.go:93]: (note I haven't actually updated other areas of the codebase to use the `AccountVisible()` check so this won't present an immediate issue)

File: internal/db/bundb/relationship.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/relationship.go]: I updated it to use the bun methodology. Buuuuut I'll be honest I'm a bit clueless on getting the tests working on postgres... :see_no_evil:

[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/db/bundb/relationship.go]: Oh yeah, for running individual tests against postgres (which is a much easier way to do it), follow step 1 but then run something like this from your command line (this is for the syslog test specifically, just replace it with whichever test you want to run):

```
go test ./internal/log -run '^TestSyslogTestSuite$' -testify.m '^(TestSyslog)$' github.com/superseriousbusiness/gotosocial/internal/log
```

File: internal/visibility/account.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/visibility/account.go:93]: A note would be fine, just to explain when you should / shouldn't use this function.

File: internal/db/bundb/relationship.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (tsmethurst) [reply internal/db/bundb/relationship.go]: I ran `GTS_DB_TYPE=postgres GTS_DB_ADDRESS=localhost go test ./internal/db/bundb -timeout 30s -run ^TestRelationshipTestSuite$ github.com/superseriousbusiness/gotosocial/internal/db/bundb` and it passed, so we're good! I'll mark this as resolved :)
